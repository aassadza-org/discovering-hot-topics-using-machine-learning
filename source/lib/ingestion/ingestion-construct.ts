#!/usr/bin/env node
/**********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/


import * as events from '@aws-cdk/aws-events';
import * as lambda from '@aws-cdk/aws-lambda';
import * as s3 from '@aws-cdk/aws-s3';
import * as cdk from '@aws-cdk/core';
import { FeedConsumer } from './feed-consumer-construct';
import { NewsCatcher } from './newscatcher-stack';
import { TwitterSearchIngestion } from './twitter-search-stack';

export interface IngestionProps {
    readonly ingestFrequency?: cdk.CfnParameter;
    readonly twitterQueryParameter?: cdk.CfnParameter;
    readonly supportedLang?: cdk.CfnParameter; //comma separated values of language codes
    readonly credentialKeyPath?: cdk.CfnParameter;
    readonly s3LoggingBucket: s3.Bucket;
    readonly rssNewsFeedQueryParameter?: cdk.CfnParameter;
    readonly rssNewsFeedConfig?: cdk.CfnParameter;
    readonly rssNewsFeedIngestFreq?: cdk.CfnParameter;
    readonly deployTwitter: cdk.CfnParameter;
    readonly deployRSSNewsFeeds: cdk.CfnParameter;
}

export class Ingestion extends cdk.Construct {
    private readonly _consumerLambdaFunc;

    constructor(scope: cdk.Construct, id: string, props: IngestionProps) {
        super(scope, id);

        const _feedConsumerlambda = new FeedConsumer(this, 'FeedConsumer', {
            functionProps: {
                runtime: lambda.Runtime.NODEJS_14_X,
                handler: 'index.handler',
                code: lambda.Code.fromAsset('lambda/ingestion-consumer'),
                timeout: cdk.Duration.minutes(5),
                reservedConcurrentExecutions: 1 // adding to throttle submissions to step functions
            },
            batchSize: 5,
            shardCount: 1,
        });
        this._consumerLambdaFunc = _feedConsumerlambda.lambdaFunction;

        // creating a common event bus messaging backbone across different ingestion sources
        const _eventBus = new events.EventBus(this, 'Bus');

        const _twitterSearch = new TwitterSearchIngestion(this, 'TwitterSearch', {
            parameters: {
                "SupportedLang": props.supportedLang!.valueAsString,
                "QueryParameter": props.twitterQueryParameter!.valueAsString,
                "SSMPathForCredentials": props.credentialKeyPath!.valueAsString,
                "IngestQueryFrequency": props.ingestFrequency!.valueAsString,
                "StreamARN": _feedConsumerlambda.kinesisStream.streamArn
            }
        });
        _twitterSearch.nestedStackResource!.addMetadata('nestedStackFileName', _twitterSearch.templateFile.slice(0, -5));

        const _deployTwitterIngestionCondition = new cdk.CfnCondition(this, 'DeployTwitterIngestion', {
            expression: cdk.Fn.conditionAnd(
                cdk.Fn.conditionEquals(props.deployTwitter, "Yes"),
                cdk.Fn.conditionNot(cdk.Fn.conditionEquals(props.twitterQueryParameter, "")),
                cdk.Fn.conditionNot(cdk.Fn.conditionEquals(props.ingestFrequency, "")),
                cdk.Fn.conditionNot(cdk.Fn.conditionEquals(props.credentialKeyPath, ""))
        )});
        // TODO - This is a temporary patch for Sonarqube, it misinterprets the use of '?' here as a control statement
        _twitterSearch.nestedStackResource?.addOverride('Condition', _deployTwitterIngestionCondition.logicalId); // NOSONAR - Rule Non-empty statements should change control flow or have at least one side-effect

        const _newsCatcher = new NewsCatcher(this, 'NewsCatcher', {
            parameters: {
                "EventBus": _eventBus.eventBusArn,
                "StreamARN": _feedConsumerlambda.kinesisStream.streamArn,
                // if the search query is blank it acts as wild char search ('*'). All feeds no filters
                "NewsSearchQuery": props.rssNewsFeedQueryParameter!.valueAsString? props.rssNewsFeedQueryParameter!.valueAsString : '',
                "Config": props.rssNewsFeedConfig!.valueAsString,
                "IngestFrequency": props.rssNewsFeedIngestFreq!.valueAsString
            }
        });
        _newsCatcher.nestedStackResource!.addMetadata('nestedStackFileName', _newsCatcher.templateFile.slice(0, -5));

        const _deployRSSFeedsIngestionCondition = new cdk.CfnCondition(this, 'DeployRSSFeeds', {
            expression: cdk.Fn.conditionAnd(
                cdk.Fn.conditionEquals(props.deployRSSNewsFeeds, "Yes"),
                cdk.Fn.conditionNot(cdk.Fn.conditionEquals(props.rssNewsFeedIngestFreq, ""))
            )});
        // TODO - This is a temporary patch for Sonarqube, it misinterprets the use of '?' here as a control statement
        _newsCatcher.nestedStackResource?.addOverride('Condition', _deployRSSFeedsIngestionCondition.logicalId); // NOSONAR - Rule Non-empty statements should change control flow or have at least one side-effect
    }

    public get consumerLambdaFunc(): lambda.Function {
        return this._consumerLambdaFunc
    }
}
