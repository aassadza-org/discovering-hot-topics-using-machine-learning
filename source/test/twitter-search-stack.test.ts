/**********************************************************************************************************************
 *  Copyright 2020-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                      *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { SynthUtils } from '@aws-cdk/assert';
import * as events from '@aws-cdk/aws-events';
import * as cdk from '@aws-cdk/core';
import { TwitterSearchIngestion } from '../lib/ingestion/twitter-search-stack';

test ('Event Bus creation', () => {
    const stack = new cdk.Stack();

    const _twtitterSearchIngestion = new TwitterSearchIngestion(stack, 'TwitterSearch', {
        parameters: {
            'EventBus': new events.EventBus(stack, 'testBus').eventBusArn,
            'SuppertedLang': "en,es,fr",
            'QueryParameter': "fakequery",
            'SSMPathForCredentials': "/discovering-hot-topics-using-machine-learning/discovering-hot-topics-using-machine-learning/twitter",
            'IngestQueryFrequency': 'cron(0 5 * * ? *)'
        }
    });

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
