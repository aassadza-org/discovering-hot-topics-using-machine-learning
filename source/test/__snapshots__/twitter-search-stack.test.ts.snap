// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event Bus creation 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters264a554884ca03dba7dd0c9045b116252a9f53b56027cfd8b337d609b3e23fb0ArtifactHash61FD114B": Object {
      "Description": "Artifact hash for asset \\"264a554884ca03dba7dd0c9045b116252a9f53b56027cfd8b337d609b3e23fb0\\"",
      "Type": "String",
    },
    "AssetParameters264a554884ca03dba7dd0c9045b116252a9f53b56027cfd8b337d609b3e23fb0S3Bucket71199E81": Object {
      "Description": "S3 bucket for asset \\"264a554884ca03dba7dd0c9045b116252a9f53b56027cfd8b337d609b3e23fb0\\"",
      "Type": "String",
    },
    "AssetParameters264a554884ca03dba7dd0c9045b116252a9f53b56027cfd8b337d609b3e23fb0S3VersionKey7F9EA7BA": Object {
      "Description": "S3 key for asset version \\"264a554884ca03dba7dd0c9045b116252a9f53b56027cfd8b337d609b3e23fb0\\"",
      "Type": "String",
    },
    "AssetParametersbb4eeb24bfa2713ce5b74bd6fdf977bd05f2b4cbbb445d39ad501a8bae80b3dfArtifactHash3A880940": Object {
      "Description": "Artifact hash for asset \\"bb4eeb24bfa2713ce5b74bd6fdf977bd05f2b4cbbb445d39ad501a8bae80b3df\\"",
      "Type": "String",
    },
    "AssetParametersbb4eeb24bfa2713ce5b74bd6fdf977bd05f2b4cbbb445d39ad501a8bae80b3dfS3BucketAFAB871D": Object {
      "Description": "S3 bucket for asset \\"bb4eeb24bfa2713ce5b74bd6fdf977bd05f2b4cbbb445d39ad501a8bae80b3df\\"",
      "Type": "String",
    },
    "AssetParametersbb4eeb24bfa2713ce5b74bd6fdf977bd05f2b4cbbb445d39ad501a8bae80b3dfS3VersionKey685F8A2B": Object {
      "Description": "S3 key for asset version \\"bb4eeb24bfa2713ce5b74bd6fdf977bd05f2b4cbbb445d39ad501a8bae80b3df\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "TwitterSearchNestedStackTwitterSearchNestedStackResource6905E30B": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Parameters": Object {
          "EventBus": Object {
            "Fn::GetAtt": Array [
              "testBus07EAAD3A",
              "Arn",
            ],
          },
          "IngestQueryFrequency": "cron(0 5 * * ? *)",
          "QueryParameter": "fakequery",
          "SSMPathForCredentials": "/discovering-hot-topics-using-machine-learning/discovering-hot-topics-using-machine-learning/twitter",
          "SuppertedLang": "en,es,fr",
          "referencetoAssetParameters264a554884ca03dba7dd0c9045b116252a9f53b56027cfd8b337d609b3e23fb0S3Bucket71199E81Ref": Object {
            "Ref": "AssetParameters264a554884ca03dba7dd0c9045b116252a9f53b56027cfd8b337d609b3e23fb0S3Bucket71199E81",
          },
          "referencetoAssetParameters264a554884ca03dba7dd0c9045b116252a9f53b56027cfd8b337d609b3e23fb0S3VersionKey7F9EA7BARef": Object {
            "Ref": "AssetParameters264a554884ca03dba7dd0c9045b116252a9f53b56027cfd8b337d609b3e23fb0S3VersionKey7F9EA7BA",
          },
        },
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.",
              Object {
                "Ref": "AWS::Region",
              },
              ".",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              Object {
                "Ref": "AssetParametersbb4eeb24bfa2713ce5b74bd6fdf977bd05f2b4cbbb445d39ad501a8bae80b3dfS3BucketAFAB871D",
              },
              "/",
              Object {
                "Fn::Select": Array [
                  0,
                  Object {
                    "Fn::Split": Array [
                      "||",
                      Object {
                        "Ref": "AssetParametersbb4eeb24bfa2713ce5b74bd6fdf977bd05f2b4cbbb445d39ad501a8bae80b3dfS3VersionKey685F8A2B",
                      },
                    ],
                  },
                ],
              },
              Object {
                "Fn::Select": Array [
                  1,
                  Object {
                    "Fn::Split": Array [
                      "||",
                      Object {
                        "Ref": "AssetParametersbb4eeb24bfa2713ce5b74bd6fdf977bd05f2b4cbbb445d39ad501a8bae80b3dfS3VersionKey685F8A2B",
                      },
                    ],
                  },
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "testBus07EAAD3A": Object {
      "Properties": Object {
        "Name": "testBus",
      },
      "Type": "AWS::Events::EventBus",
    },
  },
}
`;
